eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/LocationPermissionProxy.java

```diff
@@ -0,0 +1,151 @@
+
+
+/**
+ * A {@link SearchModule} control proxy to use before {@link CurrentLocationModule} that handles the Location permission, i.e. doesn't let queries through until
+ * the permission is granted.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class LocationPermissionProxy implements SearchModule, SinglePermission.PermissionGrantListener
+{
+    public static SearchModuleFactory<GeoPlace> factory(final SearchModuleFactory<GeoPlace> currentLocationModuleFactory)
+    {
+        return new SearchModuleFactory<GeoPlace>()
+        {
+            @Override
+            public SearchModule create(Activity activity, ResultUpdateListener<ListItem> updateListener, ItemChosenAction<GeoPlace> itemChosenAction)
+            {
+                SinglePermission locationPermission = new BroadcastReceiverPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION);
+
+                SearchModule currentLocationModule = currentLocationModuleFactory.create(activity, updateListener, itemChosenAction);
+
+                return new LocationPermissionProxy(currentLocationModule, updateListener, locationPermission,
+                        new BasicStrings(activity));
+            }
+        };
+    }
+
+
+    private final SearchModule mDelegate;
+    private final ResultUpdateListener<ListItem> mUpdateListener;
+    private final SinglePermission mLocationPermission;
+    private final Strings mStrings;
+
+    private boolean mGranted;
+
+
+    public LocationPermissionProxy(SearchModule delegate,
+                                   ResultUpdateListener<ListItem> updateListener,
+                                   SinglePermission locationPermission,
+                                   Strings strings)
+    {
+        mDelegate = delegate;
+        mUpdateListener = updateListener;
+        mLocationPermission = locationPermission;
+        mStrings = strings;
+
+        mGranted = mLocationPermission.granted();
+    }
+
+
+    @Override
+    public void shutDown()
+    {
+    }
+
+
+    @Override
+    public void onSearchQueryChange(String newQuery)
+    {
+        if (mGranted)
+        {
+            mDelegate.onSearchQueryChange(newQuery);
+        }
+        else if (newQuery.isEmpty())
+        {
+            ListItem promptingItem = new ClickableMessageItem(
+                    mStrings.get(R.string.schedjoules_location_picker_current_location),
+                    new PermissionRequestOnClick());
+
+            mUpdateListener.onUpdate(new ShowSingle<>(promptingItem, ""));
+        }
+        else
+        {
+            mUpdateListener.onUpdate(new Clear<ListItem>(newQuery));
+        }
+    }
+
+
+    /*
+     * TODO If user denies or had denied at any time the permission with "Never ask again" checked,
+     * we should not show any item (the module basically).
+     */
+
+
+    @Override
+    public void onPermissionGrantResult(boolean granted)
+    {
+        mGranted = granted;
+        if (mGranted)
```

Comment:
This should not depend on the granted state. Ideally this is determined by [shouldShowRequestPermissionRationale](https://developer.android.com/reference/android/support/v4/app/ActivityCompat.html#shouldShowRequestPermissionRationale(android.app.Activity,%20java.lang.String)), also see this comment: http://stackoverflow.com/questions/30719047/android-m-check-runtime-permission-how-to-determine-if-the-user-checked-nev#comment53800740_31925748